components/masking_functions/src/masking_functions/dictionary_flusher_thread.cpp

```diff
@@ -0,0 +1,365 @@
+    thread_attributes attributes{};
+    attributes.make_joinable();
+
+    mysql_thread_register(category_name, &psi_thread_info_, 1U);
+
+    if (mysql_thread_create(psi_thread_key_, &handle_,
+                            &attributes.get_underlying(), raw_handler,
+                            this) != 0) {
+      throw std::runtime_error{"Cannot create jthread"};
+    }
+  }
+
+  jthread(const jthread &) = delete;
+  jthread &operator=(const jthread &) = delete;
+  jthread(jthread &&) = delete;
+  jthread &operator=(jthread &&) = delete;
+  ~jthread() { my_thread_join(&handle_, nullptr); }
+
+ private:
+  handler_function_type handler_function_;
+
+  PSI_thread_key psi_thread_key_;
+  PSI_thread_info psi_thread_info_;
+
+  my_thread_handle handle_;
+
+  static void *raw_handler(void *arg) {
+    try {
+      thread_handler_context handler_ctx{};
+      const auto *self = static_cast<jthread *>(arg);
+      (self->handler_function_)();
+    } catch (const std::exception &e) {
+      std::string message{"Exception caught in jthread handler - "};
+      message += e.what();
+      LogComponentErr(ERROR_LEVEL, ER_LOG_PRINTF_MSG, message.c_str());
+    } catch (...) {
+      LogComponentErr(ERROR_LEVEL, ER_LOG_PRINTF_MSG,
+                      "Unexpected exception caught in jthread handler");
+    }
+    return nullptr;
+  }
+};
+
+constexpr char flusher_psi_category_name[]{"masking_functions"};
+constexpr char flusher_psi_thread_info_name[]{"masking_functions_dict_flusher"};
+constexpr char flusher_psi_thread_info_os_name[]{"mf_flusher"};
+
+}  // anonymous namespace
+
+namespace masking_functions {
+
+enum class dictionary_flusher_thread::state_type : std::uint8_t {
+  initial,
+  initialization_failure,
+  operational,
+  stopped
+};
+
+void dictionary_flusher_thread::jthread_deleter::operator()(
+    void *ptr) const noexcept {
+  std::default_delete<jthread>{}(static_cast<jthread *>(ptr));
+}
+
+dictionary_flusher_thread::dictionary_flusher_thread(
+    const term_cache_core_ptr &cache_core, std::uint64_t flush_interval_seconds)
+    : cache_core_{cache_core},
+      flush_interval_seconds_{flush_interval_seconds},
+      state_{state_type::initial},
+      thread_impl_{new jthread{
+          [this]() { do_periodic_reload(); }, flusher_psi_category_name,
+          flusher_psi_thread_info_name, flusher_psi_thread_info_os_name}} {
+  // we do not try to populate dict_cache here immediately as this constructor
+  // is called from the component initialization method and any call to
+  // mysql_command_query service may mess up with current THD
+
+  // the cache will be loaded during the first call to one of the dictionary
+  // functions or by the flusher thread
+}
+
+dictionary_flusher_thread::~dictionary_flusher_thread() {
+  request_termination();
+  // thread_impl_'s destructor will join the thread
+}
+
+bool dictionary_flusher_thread::request_termination() {
+  return state_.exchange(state_type::stopped) != state_type::initial;
```

Comment:
Perhaps it is better to use compare-and-swap operation instead? Failing termination request with a side-effect of still setting state_ to "stopped" state makes me uneasy, even though it might not have any practical consequences...